### ISSUE TITLE
🚀 Feature: add new `compatible-engines` rule

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [michaelfaith/eslint-plugin-package-json#48]
Splitting out from https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/43#issuecomment-1727005424 & #42: `eslint-plugin-node-dependencies` has a `compat-engines` rule. Let's merge it into this package!

https://ota-meshi.github.io/eslint-plugin-node-dependencies/rules/compat-engines

Our version of the rule should be called `compatible-engines`.  

For the most part the rule should behave the same as what's described above.  The only difference is that it shouldn't flag incompatibilities with `devDependencies` by default.

**Comment by michaelfaith** [comment]
I'm curious what this rule would add to the picture when package manager's will already provide feedback (or error if strict is enabled) when the engines conditions aren't met?

**Comment by JoshuaKGoldberg** [comment]
I like having this and other rules act as a "first round" of warnings before you have to run the package manager. This way you get feedback in the editor before having to go through a full installation attempt.

**Comment by JoshuaKGoldberg** [comment]
I like having this and other rules act as a "first round" of warnings before you have to run the package manager. This way you get feedback in the editor before having to go through a full installation attempt.

**Comment by lishaduck** [comment]
> I'm curious what this rule would add to the picture when package manager's will already provide feedback (or error if strict is enabled) when the engines conditions aren't met?

The biggest issue is that the package manager will only enforce the current node version.
So if you develop on Node 24, but support back to, say, Node 16, you won't get warnings (unless you use `nodeVersion` with pnpm).

**Comment by andreww2012** [comment]
Hey @michaelfaith, is this still relevant? I'd like to work on this one (ideally, I'd like all `eslint-plugin-node-dependencies` rules to be ported over here in some form!)

Additionally, I've [proposed some changes to the original rule](https://github.com/ota-meshi/eslint-plugin-node-dependencies/issues/275) and would appreciate some feedback on that 🙂

<details>
<summary>
Proposed changes (copied)
</summary>

Expanding on https://github.com/ota-meshi/eslint-plugin-node-dependencies/issues/248, I think `compat-engines` should actually respect the `devEngines.runtime` field by default, and fall back to `engines` only when instructed.

Since this field was not previously supported, this doesn't sound like a breaking change to me.

With that said, I'm proposing that the `devDependencies` rule option accept the following values:
- None or `true`: `devDependencies` are checked by the rule <=> an entry with `"name": "node"` exists in `devEngines.runtime`;
- `false`: `devDependencies` are not checked;
- `'engines'` (or maybe `'respectEngines'`?): `devDependencies` are checked <=> either the `engines` or `devEngines.runtime` (with a `node` entry) field exists, with the latter taking precedence;
- `{runtime: string}`: `devDependencies` are checked <=> entry with `"name": "<string passed to runtime>"` exists in `devEngines.runtime`.

### Other possible improvements

Support even more properties in the object passed to `devDependencies`:

```ts
type Options = {
  runtime?: string;
  engines?: boolean; // Also respect the `engines` field

  // Same meaning as the base rule options, but `devDependencies`-specific
  deep?: boolean;
  comparisonType?: 'normal' | 'major';
};
```

(if no properties are passed, treat it as if `true` was passed)
</details>

**Comment by michaelfaith** [comment]
> is this still relevant?

Good question.  I _think_ so.  I've been dragging my feet on this rule, because of how it sort of breaks the self-contained analysis paradigm of ESLint.  As in, the state of linting the `package.json` file includes more than just what's in the `package.json` file (and actually the package's lock file is more representative of the state for this proposal).  So, because of that it wouldn't work super great with ESLint's cache.  You could get a violation from this rule, run `pnpm dedupe`, which brings in a new set of deps that might clear the violation, but because the `package.json` didn't change, it doesn't break the cache and re-evaluate.  

I like the _idea_ of the rule, but don't like that aspect of it, and so it's had me conflicted about whether or not to implement.  With that said, I'd be open to a PR for it, if it's something you feel strongly about. 👍 I'd rather use the full word `compatible` in the name, though.  I would also want to make sure it was performant.  This is the kind of rule that has the potential of being a performance drag.  So I'd want it to go through some amount of profiling.

> Additionally, I've https://github.com/ota-meshi/eslint-plugin-node-dependencies/issues/275 and would appreciate some feedback on that 🙂

I do like the idea.  It does seem like it would be useful in the same way `compat-engines` could be.  I would suggest the following adjustments to the proposal:
- when checking `runtime`, only report if `onFail` is `error`.  If `onFail` is `download` (I think that's PNPM only), that should work as designed and the correct runtime will be downloaded if not present (no error).  Similarly `warn` or `ignore` aren't error states.
- it seems like we're just ignoring all of the other properties that are possible under `devEngines`?  It may be worth explicitly calling out that those are out of scope.
- you mention a `devDependencies` option that's boolean, but then have an options object.  Is it boolean or options object?  That's not super clear to me.
- i personally don't see the value of the `deep` property, both in terms of this proposal and the original version of the rule.  If it's not doing a deep check, the rule seems considerably less valuable.  I would drop that entirely and have it be the only behavior.
- this feels like an entirely new rule to me, rather than tacking it on to `compat-engines`.  e.g. `compatible-devEngines`  What do you think?

**Comment by andreww2012** [comment]
> <...> it sort of breaks the self-contained analysis paradigm of ESLint. As in, the state of linting the `package.json` file includes more than just what's in the `package.json` file (and actually the package's lock file is more representative of the state for this proposal). So, because of that it wouldn't work super great with ESLint's cache. You could get a violation from this rule, run `pnpm dedupe`, which brings in a new set of deps that might clear the violation, but because the `package.json` didn't change, it doesn't break the cache and re-evaluate.

While that's true, I can't see this happening often - in fact, I'd categorize this scenario as an edge case, so documenting this behavior at the rule and plugin levels should be enough IMHO.

Additionally, we could introduce a rule option (or a new plugin setting) that would take an array of arbitrary strings representing the cache key of the lockfile(s). But the proper calculation would fall entirely on the user's shoulders. (Ideally, this kind of feature should be natively supported by ESLint). I *think* this should work since ESLint invalidates its cache when the config changes.

> I'd rather use the full word `compatible` in the name, though.

Agreed 👍

> when checking `runtime`, only report if `onFail` is `error`.  If `onFail` is `download` (I think that's PNPM only), that should work as designed and the correct runtime will be downloaded if not present (no error).  Similarly `warn` or `ignore` aren't error states.

I mostly understand the reasoning behind this and agree, but just want to make sure we're on the same page:
- If the value of `onFail` is not `error`, we cannot be sure that the project is supposed to be run *only* on this runtime?
- In case of the multiple `runtime`s, we should pick the first one whose name matches and that has `onFail` set to `error`?

> you mention a `devDependencies` option that's boolean, but then have an options object.  Is it boolean or options object?  That's not super clear to me.

Either a boolean or an object. But if we separate this part of the rule into its own `compatible-devEngines` rule (which I like 👍), the boolean values would not be necessary - `false` would be the same as turning the rule off, and `true` the same as enabling it without configuring the options.

> i personally don't see the value of the `deep` property, both in terms of this proposal and the original version of the rule.  If it's not doing a deep check, the rule seems considerably less valuable. I would drop that entirely and have it be the default behavior.

My concern is, if the dependency B of the top-level dependency A fails the engines constraint, is it really our fault and not A's? Even if A specified the range for B, B got updated and changed its `engines`, is this something we can control? Of course, we could've overridden B, but that is rare and already an explicit (and dangerous) intent.

I might also add that I have been using the original rule on a number of projects and most of deep checks reports were false positives - everything continued to work just fine.

**Comment by michaelfaith** [comment]
> in fact, I'd categorize this scenario as an edge case, so documenting this behavior at the rule and plugin levels should be enough IMHO.

👍 aligned

> mostly understand the reasoning behind this and agree, but just want to make sure we're on the same page:
> 
> * If the value of `onFail` is not `error`, we cannot be sure that the project is supposed to be run _only_ on this runtime?
> * In case of the multiple `runtime`s, we should pick the first one whose name matches and that has `onFail` set to `error`?

This makes sense to me, thanks for fleshing that out more.

> My concern is, if the dependency B of the top-level dependency A fails the engines constraint, is it really our fault and not A's?

This is a really good point, and I hadn't thought of it in that way.  Flagging issues that aren't in the user's direct control doesn't seem good.  

> I might also add that I have been using the original rule on a number of projects and most of deep checks reports were false positives - everything continued to work just fine.

Honestly this has been my experience with a lot of engine incompatibilities, regardless of whether they're direct or transitive.  Engine declarations are intentionally conservative and have to cover every api that's used.  If you're not using something that touches an API that's been removed or changed in a breaking way, then you don't always feel it.

> But if we separate this part of the rule into its own `compatible-devEngines` rule (which I like 👍)...

Sounds like we're aligned on having this be a separate rule.  Would you be open to opening a new issue for this and recapping what we agreed on?  I'll update this issue to reflect the relevant bits to this rule. (namely the preference in name `compatible-engines`)

**Comment by michaelfaith** [comment]
I have good news and bad news. 

**Good news:** I've implemented an MVP for this rule: https://github.com/michaelfaith/eslint-plugin-package-json/pull/2113, and it works reasonably well.

**Bad news:** After putting in the work to implement it, I've become convinced that this rule is not a good fit for the plugin.  It adds significantly more complexity than I'm comfortable with (including an async worker, a caching mechanism, and four net new dependencies). That's especially hard to swallow when it's all just in support of one (maybe two) rules.   [I mentioned](https://github.com/michaelfaith/eslint-plugin-package-json/issues/48#issuecomment-4378413443) being conflicted about this rule, but was willing to give it a shot.  Now that I've implemented it, it reinforced those reservations.   I'd like to keep this plugin fast and stateless. Closing this and #1786 out as not planned.  

I do think there's value in this functionality existing in some form though.  Ota's is pretty solid, but is misaligned in the `devEngines` + `devDependencies` piece.  My recommendation would be to try and close the gap on that implementation.  Another idea, which I'd be happy to build at some point, once I've knocked a couple other projects out if there's interest, is to create a dedicated cli tool outside of ESLint, that does this validation.  Rather than trying to fit a square peg into a round eslint-shaped hole 😅
